lib/puppet/provider/service/systemd.rb

```diff
@@ -129,13 +129,25 @@ def unmask
     systemctl_change_enable(:unmask)
   end
 
-  def restartcmd
-    [command(:systemctl), "restart", @resource[:name]]
+  def restart
+    cmds = []
+    cmds << :systemctl << 'restart' << @resource[:name]
+    begin
+      execute(cmds)
+    rescue Puppet::ExecutionFailure
+      raise Puppet::Error.new(prepare_error_message(@resource[:name], 'restart'), $!)
```

Comment:
Prefer `$ERROR_STATUS` over `$!` with the [English](http://ruby-doc.org/stdlib-2.0.0/libdoc/English/rdoc/English.html) module; I've forgotten as much Perl as I can and can't recall what the magic variables mean anymore :)